catalog/app/containers/Bucket/Queries/requests/athena.ts

```diff
@@ -0,0 +1,334 @@
+  id?: string
+  outputBucket?: string
+  query?: string
+  status?: string // 'QUEUED' | 'RUNNING' | 'SUCCEEDED' | 'FAILED' | 'CANCELLED'
+  workgroup?: Athena.WorkGroupName
+}
+
+export interface QueryExecutionsResponse {
+  list: QueryExecution[]
+  next?: string
+}
+
+interface QueryExecutionsArgs {
+  athena: Athena
+  prev: QueryExecutionsResponse | null
+  workgroup: string
+}
+
+function parseQueryExecution(queryExecution: Athena.QueryExecution): QueryExecution {
+  return {
+    catalog: queryExecution?.QueryExecutionContext?.Catalog,
+    completed: queryExecution?.Status?.CompletionDateTime,
+    created: queryExecution?.Status?.SubmissionDateTime,
+    db: queryExecution?.QueryExecutionContext?.Database,
+    id: queryExecution?.QueryExecutionId,
+    outputBucket: queryExecution?.ResultConfiguration?.OutputLocation,
+    query: queryExecution?.Query,
+    status: queryExecution?.Status?.State,
+    workgroup: queryExecution?.WorkGroup,
+  }
+}
+
+async function fetchQueryExecutions({
+  athena,
+  prev,
+  workgroup,
+}: QueryExecutionsArgs): Promise<QueryExecutionsResponse> {
+  try {
+    const executionIdsOutput = await athena
+      .listQueryExecutions({ WorkGroup: workgroup, NextToken: prev?.next })
+      .promise()
+
+    const ids = executionIdsOutput.QueryExecutionIds
+    if (!ids || !ids.length)
+      return {
+        list: [],
+        next: executionIdsOutput.NextToken,
+      }
+
+    const executionsOutput = await athena
+      ?.batchGetQueryExecution({ QueryExecutionIds: ids })
+      .promise()
+    const parsed = (executionsOutput.QueryExecutions || []).map(parseQueryExecution)
+    const list = (prev?.list || []).concat(parsed)
+    return {
+      list,
+      next: executionIdsOutput.NextToken,
+    }
+  } catch (e) {
+    // eslint-disable-next-line no-console
+    console.log('Unable to fetch')
+    // eslint-disable-next-line no-console
+    console.error(e)
+    throw e
+  }
+}
+
+export function useQueryExecutions(
+  workgroup: string,
+  prev: QueryExecutionsResponse | null,
+): AsyncData<QueryExecutionsResponse> {
+  const athena = AWS.Athena.use()
+  return useData(
+    fetchQueryExecutions,
+    { athena, prev, workgroup },
+    { noAutoFetch: !workgroup },
+  )
+}
+
+async function waitForQueryStatus(
+  athena: Athena,
+  QueryExecutionId: string,
+): Promise<Athena.QueryExecution | null> {
+  let status: string | undefined = 'QUEUED'
+  let queryExecution = null
+  while (status === 'QUEUED' || status === 'RUNNING') {
+    // NOTE: await is used to intetionally pause loop and make requests in series
+    // eslint-disable-next-line no-await-in-loop
+    const statusData = await athena.getQueryExecution({ QueryExecutionId }).promise()
+    status = statusData?.QueryExecution?.Status?.State
+    queryExecution = statusData?.QueryExecution || null
+    if (status === 'SUCCEEDED') {
+      return queryExecution
+    }
+    // eslint-disable-next-line no-await-in-loop
+    await wait(1000)
```

Comment:
seems like this will wait a second before throwing an error, so we probably need to move throw statement inside the loop as well under some condition. and after this we wont need to check for any conditions in the "while" condition clause (making it `while (true) {`). after this, we wont need `status` and `queryExecution` variables outside of "while", effectively transforming this function into the variant i've posted some time ago).
also, i dont see how the case with `status === undefined` is handled

Reply:
I use you variant now